src/overlays/actors/ovl_En_Door/z_en_door.c

```diff
@@ -27,29 +40,342 @@ const ActorInit En_Door_InitVars = {
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Door/func_809FC5D0.s")
+void EnDoor_AjarClose(EnDoor* this, GlobalContext* globalCtx) {
+    if (Math_ApproxUpdateScaledS(&this->actor.posRot.rot.y, 0, 0x700)) {
+        this->actionFunc = EnDoor_Idle;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Door/func_809FC878.s")
+void EnDoor_Open(EnDoor* this, GlobalContext* globalCtx) {
+    s32 i;
+    s32 numEffects;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Door/func_809FC8C0.s")
+    if (DECR(this->lockTimer) == 0) { // Wait for the lock to finish if applicable
+        if (SkelAnime_FrameUpdateMatrix(&this->skelAnime)) {
+            // If done, return to the idle action and free the player
+            this->actionFunc = EnDoor_Idle;
+            this->unk_191 = 0;
+        } else if (func_800A56C8(&this->skelAnime, sDoorAnimOpenFrames[this->unk_190])) {
+            // Play the appropriate sound on open
+            Audio_PlayActorSound2(&this->actor,
+                                  (globalCtx->sceneNum == SCENE_HAKADAN || globalCtx->sceneNum == SCENE_HAKADANCH ||
+                                   globalCtx->sceneNum == SCENE_HIDAN)
+                                      ? NA_SE_EV_IRON_DOOR_OPEN
+                                      : NA_SE_OC_DOOR_OPEN);
+            if (this->skelAnime.animPlaybackSpeed < 1.5f) {
+                numEffects = (s32)(Math_Rand_ZeroOne() * 30.0f) + 50;
+                for (i = 0; i < numEffects; i++) {
+                    EffectSsBubble_Spawn(globalCtx, &this->actor.posRot.pos, 60.0f, 100.0f, 50.0f, 0.15f);
+                }
+            }
+        } else if (func_800A56C8(&this->skelAnime, sDoorAnimCloseFrames[this->unk_190])) {
+            // Play the appropriate sound on close
+            Audio_PlayActorSound2(&this->actor,
+                                  (globalCtx->sceneNum == SCENE_HAKADAN || globalCtx->sceneNum == SCENE_HAKADANCH ||
+                                   globalCtx->sceneNum == SCENE_HIDAN)
+                                      ? NA_SE_EV_IRON_DOOR_CLOSE
+                                      : NA_SE_EV_DOOR_CLOSE);
+        }
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Door/func_809FC8F4.s")
+void EnDoor_Update(Actor* thisx, GlobalContext* globalCtx) {
+    EnDoor* this = THIS;
+    this->actionFunc(this, globalCtx);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Door/func_809FC928.s")
+s32 EnDoor_OverrideLimbDraw(GlobalContext* globalCtx, s32 limbIndex, Gfx** dList, Vec3f* pos, Vec3s* rot,
+                            Actor* thisx) {
+    s32 pad;
+    TransitionActorEntry* transitionEntry;
+    Gfx** temp_a2;
+    s32 pad2;
+    s16 phi_v0_2;
+    s32 phi_v0;
+    EnDoor* this = THIS;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Door/func_809FC998.s")
+    if (limbIndex == 4) {
+        temp_a2 = &D_809FCEE4[(s8)this->displaylistIdx];
```

Comment:
```suggestion
        temp_a2 = D_809FCEE4[(s8)this->displaylistIdx];
```
this & is unneeded and causes a type warning. Also is the (s8) cast needed? Seem to be a lot of stray casts around.